### PR TITLE
Expose code linters as a LSP plugin

### PR DESCRIPTION
**Pull request by vsevolodstep-db**
## Changes
Add a PyLSP plugin entrypoint for UCX code linters. When a user has `python-lsp-server` installed, UCX will automatically register an LSP plugin on installation. 

### Tests

- [x] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/lsp_plugin.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from pylsp import hookimpl  # type: ignore
+from pylsp.workspace import Document, Workspace  # type: ignore
+from databricks.sdk.service.workspace import Language
+
+from databricks.labs.ucx.source_code.linters.context import LinterContext
+from databricks.labs.ucx.source_code.lsp import Diagnostic
+
+# TODO: initialize migration index and session state from config / env variables
+languages = LinterContext(index=None, session_state=None)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/lsp_plugin.py:9]: Init linter context for every document, because it's stateful

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/linters/context.py [hunk 2]
```diff
@@ -12,26 +12,36 @@
...
+            DbutilsLinter(session_state),
+        ]
+        python_fixers: list[Fixer] = []
+
+        sql_linters: list[Linter] = [FromDbfsFolder()]
+        sql_fixers: list[Fixer] = []
+
+        if index is not None:
+            from_table = FromTable(index, session_state=session_state)
+            python_linters.append(SparkSql(from_table, index, session_state))
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/linters/context.py:32]: This is now the last linter, where it was the first before

File: src/databricks/labs/ucx/source_code/linters/context.py [hunk 3]
```diff
@@ -12,26 +12,36 @@
 
 
 class LinterContext:
-    def __init__(self, index: MigrationIndex, session_state: CurrentSessionState | None = None):
+    def __init__(self, index: MigrationIndex | None, session_state: CurrentSessionState | None = None):
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/source_code/linters/context.py:15]:
```suggestion
    def __init__(self, index: MigrationIndex | None, session_state: CurrentSessionState | None = None):
```
```suggestion
    def __init__(self, index: MigrationIndex | None = None, session_state: CurrentSessionState | None = None):
```
```suggestion
    def __init__(self, index: MigrationIndex | None, session_state: CurrentSessionState | None = None):
```

**Comment by vsevolodstep-db** [comment]
Not sure what can I do about no-check check there. LSP plugin system requires `pylsp_lint` hook to have two arguments with specified names (`workspace` and `document`)

File: tests/unit/source_code/test_lsp_plugin.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+def temp_document(doc_text, ws):
+    with tempfile.NamedTemporaryFile(mode="w", dir=ws.root_path, delete=False) as temp_file:
+        name = temp_file.name
+        temp_file.write(doc_text)
+    doc = Document(uris.from_fs_path(name), ws)
+    return name, doc
+
+
+def test_pylsp_lint(workspace):
+    code = 'sc.emptyRDD()'
```
Comment (nfx) [inline tests/unit/source_code/test_lsp_plugin.py:30]: we may want to add one more test with 2-3 lines of code to make sure line offsets are working fine.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**